### DISCUSSION TITLE
Question: Set jobdatamap value

### DISCUSSION THREAD

**Discussion by Sicos1977** [quartznet/quartznet#904] [General] [answered]
Is there a way to set a jobdatamap value from the scheduler in a job that is active in the scheduler. I tried the code below but that does not set the jobdatamap in the job itself. 

```c#
            var jobDetail = Scheduler.GetJobDetail(jobKey).Result;
            jobDetail.JobDataMap.Put("ManualPaused", user?.DisplayName);
            Scheduler.PauseJob(jobKey);
```

**Comment by lahma** [comment, marked as answer]
There's overload in `IScheduler`:

``` csharp
/// <summary>
/// Add the given <see cref="IJob" /> to the Scheduler - with no associated
/// <see cref="ITrigger" />. The <see cref="IJob" /> will be 'dormant' until
/// it is scheduled with a <see cref="ITrigger" />, or <see cref="TriggerJob(Quartz.JobKey, CancellationToken)" />
/// is called for it.
/// </summary>
/// <remarks>
/// With the <paramref name="storeNonDurableWhileAwaitingScheduling"/> parameter
/// set to <code>true</code>, a non-durable job can be stored.  Once it is
/// scheduled, it will resume normal non-durable behavior (i.e. be deleted
/// once there are no remaining associated triggers).
/// </remarks>
Task AddJob(
    IJobDetail jobDetail, 
    bool replace, 
    bool storeNonDurableWhileAwaitingScheduling,
    CancellationToken cancellationToken = default);
```

So calling this with

`await scheduler.AddJob(jobDetail, replace: true, storeNonDurableWhileAwaitingScheduling: true);` 

should do the trick.

**Comment by Sicos1977** [comment]
Will that just replace the job that is already present and use the triggers that are added to that job?

**Comment by lahma** [comment]
If you follow down to the rabbit hole, you'll end up here (I'm using persistent store as an example):

https://github.com/quartznet/quartznet/blob/f3731a7c0ef4e1226f35d53531432f42606a86df/src/Quartz/Impl/AdoJobStore/JobStoreSupport.cs#L898-L928

Which effectively will just update the job detail data:

https://github.com/quartznet/quartznet/blob/f3731a7c0ef4e1226f35d53531432f42606a86df/src/Quartz/Impl/AdoJobStore/StdAdoDelegate.cs#L734-L758

So the job will just be updated and and triggers still refer the same job key. Job is retrieved for the next trigger to fire with current data.

**Comment by Sicos1977** [comment]
Thanks, going to try that out tomorrow.

**Comment by lahma** [comment]
Closing as answered.

**Comment by Sicos1977** [comment]
I tried your solution but it keeps throwing an exception. I'm trying this from a remote scheduler

![image](https://user-images.githubusercontent.com/6692947/88807351-e931e800-d1b1-11ea-91a6-b7642f348851.png)

**Comment by lahma** [comment]
Can you please paste the full exception stack trace?

**Comment by Sicos1977** [comment]
at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg) in f:\dd\ndp\clr\src\BCL\system\runtime\remoting\realproxy.cs:line 427
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type) in f:\dd\ndp\clr\src\BCL\system\runtime\remoting\realproxy.cs:line 852
   at Quartz.Simpl.IRemotableQuartzScheduler.AddJob(IJobDetail jobDetail, Boolean replace)
   at Quartz.Impl.RemoteScheduler.<>c__DisplayClass32_0.<AddJob>b__0(IRemotableQuartzScheduler x)
   at Quartz.Impl.RemoteScheduler.CallInGuard(Action`1 action)

**Comment by lahma** [comment]
What is message of exception and possible inner exception? So far I haven't seen any details that would explain the error...

**Comment by Sicos1977** [comment]
Message = "Error communicating with remote scheduler."
InnerException = {"Server encountered an internal error. For more information, turn off customErrors in the server's .config file."}

**Comment by lahma** [comment]
Maybe you need to follow that guidance, I suspect remoting won't expose too much about what is happening without you allowing it. Might be easier to start with local scheduler and make sure the concept works.

**Comment by Sicos1977** [comment]
Yeah i'll try that, hope it will give an error that makes more sence

**Comment by Sicos1977** [comment]
That seems to work without any problems

**Comment by Sicos1977** [comment]
I think I figured it out. I use the ramstore to store my jobs.

When I change the code to this then everything works as expected.

```c#
            var jobDetail = Scheduler.GetJobDetail(jobKey).Result;
            jobDetail.JobDataMap.Put("ManualPaused", "Test");
            Scheduler.PauseJob(jobKey);
            try
            {
                Scheduler.AddJob(jobDetail, true, true);
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw;
            }
```

I think `bool storeNonDurableWhileAwaitingScheduling` defaults to `false` when I set it to `true` then it works

**Comment by lahma** [comment]
This is why explicitly set it true in my example, it's safest for your use case 😉
